Cython/Utility/TypeConversion.c

```diff
@@ -276,30 +276,81 @@ static CYTHON_INLINE const char* __Pyx_PyUnicode_AsStringAndSize(PyObject* o, Py
 #if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII || __PYX_DEFAULT_STRING_ENCODING_IS_UTF8
     if (PyUnicode_Check(o)) {
-        return __Pyx_PyUnicode_AsStringAndSize(o, length);
+        char* result = __Pyx_PyUnicode_AsStringAndSize(o, length);
+        if (unlikely(result == NULL)) {
+            goto error;
+        } else {
+            return result;
+        }
     } else
 #endif
 
     if (PyByteArray_Check(o)) {
+        return PyErr_Warn(PyExc_RuntimeWarning,
+                          "Cython: Pointer take from resizable `bytearray` object. "
+                          "Consider using `unsigned char[::1]` instead. "
+                          "This will also ensure the underlying `bytearray` is not resized while held."
+        );
 #if (CYTHON_ASSUME_SAFE_SIZE && CYTHON_ASSUME_SAFE_MACROS) || (CYTHON_COMPILING_IN_PYPY && (defined(PyByteArray_AS_STRING) && defined(PyByteArray_GET_SIZE)))
         *length = PyByteArray_GET_SIZE(o);
         return PyByteArray_AS_STRING(o);
 #else
         *length = PyByteArray_Size(o);
-        if (*length == -1) return NULL;
+        if (*length == -1) goto error;
         return PyByteArray_AsString(o);
 #endif
 
-    } else
-    {
+    } else if (PyBytes_Check(o)) {
         char* result;
         int r = PyBytes_AsStringAndSize(o, &result, length);
         if (unlikely(r < 0)) {
-            return NULL;
+            goto error;
+        } else {
+            return result;
+        }
+    } else {
+        char* result;
+        int r;
+#if !CYTHON_COMPILING_IN_LIMITED_API || __PYX_LIMITED_VERSION_HEX >= 0x030B0000
+        // Check that `o`:
+        // * Supports the Python Buffer Protocol
+        // * There is no need to keep the buffer around (as we cannot return it)
+        PyBufferProcs *pb = Py_TYPE(o)->tp_as_buffer;
+        if (pb == NULL || pb->bf_getbuffer == NULL || pb->bf_releasebuffer != NULL) {
+            goto error;
+        }
+
+        // Try to acquire buffer from `o`
+        Py_buffer view;
+        r = PyObject_GetBuffer(o, &view, PyBUF_SIMPLE);
+        if (unlikely(r < 0)) {
+            goto error;
+        } else {
+            result = (char*) view.buf;
+            *length = view.len;
+            PyBuffer_Release(&view);  // Release to clean up buffer (decref object `o`)
+            return result;
+        }
+#else
+        // Fallback for the Limited API on Python pre-3.11
+        r = PyArg_Parse(o, "y#", &result, &length);
+        if (unlikely(r < 0)) {
+            goto error;
         } else {
             return result;
         }
+#endif
     }
+
+    error:
+        // Clear existing errors and return our standardized one
+        PyErr_Clear();
+        __Pyx_TypeName o_type_name = __Pyx_PyType_GetFullyQualifiedName(Py_TYPE(o));
+        PyErr_Format(PyExc_TypeError,
+            "a bytes-like object is required, not '" __Pyx_FMT_TYPENAME "'",
+            o_type_name);
+        *length = -1;
+        return NULL;
```

Comment:
Added a generic error for all cases to use here. Please let me know what you think about this and if there are things we want to improve about it